File: python/udp_recv.py
import socket
import struct

sock = socket.socket(socket.AF_INET, socket.SOCK_RAW, socket.IPPROTO_UDP)

print("Listening for UDP packets on port 12345.......")

while True:
    packet, addr = sock.recvfrom(65535)
    ip_header = packet[:20]
    iph = struct.unpack('!BBHHHBBH4s4s', ip_header)
    version_ihl = iph[0]
    ihl = version_ihl & 0xF
    iph_length = ihl * 4 
    protocol = iph[6]

    if protocol != socket.IPPROTO_UDP:
        continue

    udp_header = packet[iph_length:iph_length+8]
    udph = struct.unpack('!HHHH', udp_header)
    src_port = udph[0]
    dest_port = udph[1]
    udp_length = udph[2]

    if dest_port != 12345:
        continue

    data_start = iph_length + 8
    data = packet[data_start:data_start + udp_length - 8]
    print(f"Received: {data.decode('utf-8', 'ignore')}")

File: C/udp_recv.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/socket.h>
#include <netinet/in.h>
#include <netinet/ip.h>
#include <netinet/udp.h>
#include <arpa/inet.h>
#include <unistd.h>

int main() {
  int sock = socket(AF_INET, SOCK_RAW, IPPROTO_UDP);
  if (sock < 0) {
    perror("socket");
    exit(1);
  }

  char buffer[65536];
  struct sockaddr_in src;
  socklen_t src_len = sizeof(src);

  printf("Listening for UDP packets on port 12345......\n");

  while (1) {
    int len = recvfrom(sock, buffer, sizeof(buffer), 0, (struct sockaddr *)&src, &src_len);
    if (len < 0) {
      perror("recvfrom");
      continue;
    }

    struct iphdr *iph = (struct iphdr *)buffer;
    if (iph->protocol != IPPROTO_UDP) continue;

    struct udphdr *udph = (struct udphdr *)(buffer + iph->ihl * 4);
    if (ntohs(udph->dest) != 12345) continue;

    char *data = buffer + iph->ihl * 4 + sizeof(struct udphdr);
    int data_len = ntohs(udph->len) - sizeof(struct udphdr);
    data[data_len] = '\0';

    printf("Received: %s\n", data);
  }

  close(sock);
  return 0;
}
